c++/src/kj/mutex.c++

```diff
@@ -547,8 +547,15 @@ void Mutex::wait(Predicate& predicate, Maybe<Duration> timeout) {
     } else {
       DWORD error = GetLastError();
       if (error == ERROR_TIMEOUT) {
-        // Timed out. Skip predicate check.
-        return;
+        // As mentioned above, Windows sleeps are inaccurate, so we must double check endTime.
+        auto e = KJ_ASSERT_NONNULL(endTime, "unexpected timeout from SleepConditionVariableSRW()");
+        if (e <= clock->now()) {
```

Comment:
It occurs to me that we're reading `clock->now()` here and then immediately reading it again below if we continue around the loop. Is that worth avoiding?

Reply:
Maybe if there's a reasonably readable way. I suppose we could remove the check entirely and do something like:

```cpp
if (error == ERROR_TIMEOUT) {
  // We may have been woken up early, so don't return yet. Instead, proceed through
  // the loop and rely on our endTime recalculation to detect if we timed out.
} else {
  KJ_FAIL_WIN32(...);
}
```

It'd be a minor behavior change (we'd check for exceptions where before we didn't), but that seems fine.

Reply:
Amended as such, open to suggestions though.

Reply:
Oh! Yes! That's fine and much better than what I was thinking.